java/ql/lib/semmle/code/java/security/SuspiciousRegexpRangeQuery.qll

```diff
@@ -0,0 +1,302 @@
+  result.isRange(_, "~")
+  or
+  // the same with " " and "!". " " is the first printable character, and "!" is the first non-white-space printable character.
+  result.isRange([" ", "!"], _)
+  or
+  // I've seen this often enough, looks OK.
+  result.isRange("@", "_")
+  or
+  // starting from the zero byte is a good indication that it's purposely matching a large range.
+  result.isRange(0.toUnicode(), _)
+}
+
+/** Gets all chars between (and including) `low` and `high`. */
+bindingset[low, high]
+private string inRange(string low, string high) {
+  result = [toCodePoint(low) .. toCodePoint(high)].toUnicode()
+}
+
+/** A module computing an equivalent character class for an overly wide range. */
+module RangePrinter {
+  bindingset[char]
+  private string next(string char) {
+    exists(int prev, int next |
+      prev.toUnicode() = char and
+      next.toUnicode() = result and
+      next = prev + 1
+    )
+  }
+
+  bindingset[char]
+  private string prev(string char) {
+    exists(int prev, int next |
+      prev.toUnicode() = char and
+      next.toUnicode() = result and
+      next = prev - 1
+    )
+  }
+
+  /** Gets the points where the parts of the pretty printed range should be cut off. */
+  private string cutoffs() { result = ["A", "Z", "a", "z", "0", "9"] }
+
+  /** Gets the char to use in the low end of a range for a given `cut` */
+  private string lowCut(string cut) {
+    cut = ["A", "a", "0"] and
+    result = cut
+    or
+    cut = ["Z", "z", "9"] and
+    result = next(cut)
+  }
+
+  /** Gets the char to use in the high end of a range for a given `cut` */
+  private string highCut(string cut) {
+    cut = ["Z", "z", "9"] and
+    result = cut
+    or
+    cut = ["A", "a", "0"] and
+    result = prev(cut)
+  }
+
+  /** Gets the cutoff char used for a given `part` of a range when pretty-printing it. */
+  private string cutoff(OverlyWideRange range, int part) {
+    exists(int low, int high | isRange(range, low, high) |
+      result =
+        rank[part + 1](string cut |
+          cut = cutoffs() and low < toCodePoint(cut) and toCodePoint(cut) < high
+        |
+          cut order by toCodePoint(cut)
+        )
+    )
+  }
+
+  /** Gets the number of parts we should print for a given `range`. */
+  private int parts(OverlyWideRange range) { result = 1 + strictcount(cutoff(range, _)) }
+
+  /** Holds if the given part of a range should span from `low` to `high`. */
+  private predicate part(OverlyWideRange range, int part, string low, string high) {
+    // first part.
+    part = 0 and
+    (
+      range.isRange(low, high) and
+      parts(range) = 1
+      or
+      parts(range) >= 2 and
+      range.isRange(low, _) and
+      high = highCut(cutoff(range, part))
+    )
+    or
+    // middle
+    part >= 1 and
+    part < parts(range) and
+    low = lowCut(cutoff(range, part - 1)) and
+    high = highCut(cutoff(range, part))
+    or
+    // last.
+    part = parts(range) and
+    low = lowCut(cutoff(range, part - 2)) and
```

Comment:
The numbering here is a bit off (you're skipping `part = parts(range) - 1`) - it ends up working out, but best to fix it.
```suggestion
    part < parts(range) - 1 and
    low = lowCut(cutoff(range, part - 1)) and
    high = highCut(cutoff(range, part))
    or
    // last.
    part = parts(range) - 1 and
    low = lowCut(cutoff(range, part - 1)) and
```